### PR TITLE
mds: skip directory size checks for reintegration

### PR DESCRIPTION
**Pull request by batrick**
## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by neha-ojha** [comment]
jenkins test submodules

File: src/mds/Server.cc [hunk 1]
```diff
@@ -8240,10 +8240,10 @@ void Server::handle_client_rename(MDRequestRef& mdr)
     if (!check_access(mdr, destdn->get_dir()->get_inode(), MAY_WRITE))
       return;
 
-    if (!check_fragment_space(mdr, destdn->get_dir()))
+    if (!linkmerge && !check_fragment_space(mdr, destdn->get_dir()))
```
Comment (vshankar) [inline src/mds/Server.cc:8243]: I'm a bit confused about what `linkmerge` signifies. It is set when source and dest refer to the same inode. Then, there is a check later in this function that asserts if the source inode is a stray inode when `linkmerge` is set, which looks to me that this flag signifies if the current operation is a stray inode reintegration.

If that's the case, should this be::

```
if (linkmerge && ....)
```

Comment (batrick) [reply src/mds/Server.cc:8243]: > I'm a bit confused about what `linkmerge` signifies. It is set when source and dest refer to the same inode. Then, there is a check later in this function that asserts if the source inode is a stray inode when `linkmerge` is set, which looks to me that this flag signifies if the current operation is a stray inode reintegration.

Yes. Keep in mind that `rename(3)` syscall does not perform any action (surprisingly!) when `src` and `dest` are hardlinks to the same file. See also:

> If the old argument and the new argument resolve to the same existing file, rename() shall return successfully and perform no other action.

From: https://pubs.opengroup.org/onlinepubs/9699919799/

The MDS does not actually test this for this specific case -- it relies on the client to catch that and return without involving the MDS.

For this reason, the MDS assumes this is a stray reintegration whenever it sees a rename operation that does this.

> If that's the case, should this be::

No, we do not want to check fragment space if it's a `linkmerge`.

Comment (vshankar) [reply src/mds/Server.cc:8243]: > No, we do not want to check fragment space if it's a `linkmerge`.

doh! obviously. I incorrectly interpreted the condition ;P

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-Dec-22

**Comment by batrick** [comment]
Good to merge with approval.

**Review by lxbsz — APPROVED** [review]
LGTM.

**Comment by batrick** [comment]
Thanks @lxbsz !

**Pull request merged**